src/analysis/normed_space/spectrum.lean

```diff
@@ -0,0 +1,182 @@
+* `is_compact`: the spectrum is compact.
+* `spectral_radius_le_norm`: the spectral radius is bounded above by the norm.
+* `resolvent_has_deriv_at`: the resolvent function is differentiable on the resolvent set.
+
+
+## TODO
+
+* after we have Liouville's theorem, prove that the spectrum is nonempty when the
+  scalar field is ℂ.
+* compute all derivatives of `resolvent a`.
+
+-/
+
+/-- The *spectral radius* is the supremum of the norm of elements in the spectrum.
+    The spectrum is compact, and so when it is nonempty, this supremum can be
+    achieved. Note that it is possible for `spectrum 𝕜 a = ∅`.  In this case,
+    `spectral_radius a = 0` by the definition of `Sup` on `ℝ`.-/
+noncomputable def spectral_radius (𝕜 : Type*) {A : Type*} [normed_field 𝕜] [ring A]
+  [algebra 𝕜 A] (a : A) : ℝ :=
+Sup (norm '' (spectrum 𝕜 a))
+
+
+namespace spectrum
+
+section spectrum_compact
+
+variables {𝕜 : Type*} {A : Type*}
+variables [normed_field 𝕜] [normed_ring A] [normed_algebra 𝕜 A] [complete_space A]
+
+local notation `σ` := spectrum 𝕜
+local notation `ρ` := resolvent_set 𝕜
+local notation `↑ₐ` := algebra_map 𝕜 A
+
+lemma mem_resolvent_set_of_nearby {a : A} {k k' : 𝕜} (hk : k ∈ ρ a)
+  (hkk' : ∥k' - k∥ < ∥(↑hk.unit⁻¹ : A)∥⁻¹) :
+  k' ∈ ρ a :=
+begin
+  refine (units.unit_of_nearby hk.unit (↑ₐk' - a) _).is_unit,
+  calc ∥(↑ₐk' - a) - (↑ₐk - a)∥
+       = ∥↑ₐ(k' - k)∥         : by rw [ring_hom.map_sub, sub_sub_sub_cancel_right]
+  ...  = ∥k' - k∥ * ∥(1 : A)∥ : by rw [algebra.algebra_map_eq_smul_one,norm_smul]
+  ...  = ∥k' - k∥             : by simp [normed_algebra.norm_one 𝕜 A]
+  ...  < ∥↑hk.unit⁻¹∥⁻¹       : hkk',
+end
+
+lemma is_open_resolvent_set (a : A) : is_open (ρ a) :=
+begin
+  haveI := normed_algebra.nontrivial 𝕜 A,
+  apply metric.is_open_iff.mpr,
+  intros k hk,
+  refine ⟨∥↑hk.unit⁻¹∥⁻¹, inv_pos.mpr (units.norm_pos (hk.unit⁻¹)), _⟩,
+  intros k' hk',
+  rw [metric.mem_ball, dist_eq_norm] at hk',
+  exact mem_resolvent_set_of_nearby hk hk',
+end
+
+/-- The `resolvent_set` as a term of `opens 𝕜` -/
+def open_resolvent_set (a : A) : topological_space.opens 𝕜 :=
+⟨ρ a, is_open_resolvent_set a⟩
+
+lemma is_closed (a : A) : is_closed (σ a) :=
+is_open.is_closed_compl (is_open_resolvent_set a)
+
+lemma mem_resolvent_of_norm_lt {a : A} {k : 𝕜} (h : ∥a∥ < ∥k∥) :
+  k ∈ ρ a :=
+begin
+  rw [resolvent_set,set.mem_set_of_eq,algebra.algebra_map_eq_smul_one],
+  have k_pos := lt_of_le_of_lt (norm_nonneg a) h,
+  let ku := units.mk0 k (ne_zero_of_norm_pos k_pos),
+  have lt_one :=
+    calc  ∥ku⁻¹ • a∥ = ∥↑ku⁻¹ • a∥   : rfl
+      ...            = ∥(↑ku)⁻¹ • a∥ : by rw units.coe_inv' ku
+      ...            = ∥k⁻¹∥ * ∥a∥   : norm_smul k⁻¹ a
+      ...            = ∥k∥⁻¹ * ∥a∥   : by rw normed_field.norm_inv
+      ...            < 1            : (inv_mul_lt_iff k_pos).mpr (by simp [h]),
+  have : is_unit (1 - ku⁻¹ • a), from (units.one_sub (ku⁻¹ • a) lt_one).is_unit,
+  rwa ←is_unit.smul_sub_iff_sub_inv_smul at this,
```

Comment:
Maybe something like this?
```suggestion
  let ku := units.mk (↑ₐk) (↑ₐk⁻¹) sorry sorry,
  have hku : ∥-a∥ < ∥(↑ku⁻¹:A)∥⁻¹ := by simpa [ku, algebra_map_isometry] using h,
  simpa [ku, sub_eq_add_neg, algebra.algebra_map_eq_smul_one] using (ku.add (-a) hku).is_unit,
```
(`sorry`s because I don't know if there is a convenience constructor available to make `↑ₐk` a unit, maybe you've added it lately?)

Reply:
yes, I found a nice thing: `units.map (↑ₐ).to_monoid_hom (units.mk0 k hk)` where `hk : k ≠ 0`

Reply:
Nice!